# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
lambda x: f'Hello, {x}'

<function __main__.<lambda>(x)>

#### Saving to a Variable

In [4]:
#Lambda funcitons can be used as self-invoking functions
#aka functions taht immediately call themselves after being defined
#same as above but self-invoking
(lambda x: f'Hello, {x}!')('Alex')

hello = lambda x: f'Hello, {x}!'
print(hello)
hello('Alex')

<function <lambda> at 0x000001487092FEE0>


'Hello, Alex!'

#### Multiple Inputs

In [5]:
#lambda functions are just in the format
#lambda <inputs>: <return value>
#so, if we want multiple inputs, we just specify multiple inputs
addNums = lambda a,b: a+b

addNums(2,5)

7

#### Passing a Lambda into a Function

In [10]:
#most commonly done with build-ins such as map, filter, and sorted
#sorted example:

animals = [{'name': 'Fennec Fox', 'color': 'sand'}, {'name': 'orca', 'color': 'black and white'}]

print(type(animals), type(animals[0]))
#sorted(animals) sorted cannot directly handle dictionaries
#we could do that with a regular function
def whichKey(a_dict):
    return a_dict['color']

sorted(animals, key=whichKey)

#That is a very simple regular function that probably won't be used a second time
#it would be easier/function to use a lambda function in place of the regular function here
#same thing but with a lambda
sorted(animals, key=lambda x:x['color'])

<class 'list'> <class 'dict'>


[{'name': 'orca', 'color': 'black and white'},
 {'name': 'Fennec Fox', 'color': 'sand'}]

#### Returning a Lambda from a Function

In [13]:
#a type of closure - function called from within another function

#regular function example
def multiply_normal (num):
    return num*2

#3fining a function within a function is powwible
def return_a_function():
    def multiply(num):
        return num*2
    return multiply

f_return = return_a_function()
print(f_return)
print(f_return(15), '\n\n')

#somewhat similar returning a lambda function
# the goal here is to set up functions with slightly ferent behaviour
def returnLambda(a, b):
    return lambda x,y: x+y+a+b

#we will get different behaviour from the lambda function depending on how the outer function is called

#In the below example, a=600, b=40, x=400, y=60
numAdds = returnLambda(600,40)
print(numAdds)
print(numAdds(400,60))


strAdds = returnLambda('cat', 'dog')
print(strAdds)
print(strAdds('shark', 'giraffe'))

<function return_a_function.<locals>.multiply at 0x0000014870A168B0>
30 


<function returnLambda.<locals>.<lambda> at 0x0000014870A165E0>
1100
<function returnLambda.<locals>.<lambda> at 0x0000014870A241F0>
sharkgiraffecatdog


#### If Statements within Lambdas

In [ ]:
#if statements within Lambdas functione xactly the same as if statements within a return statement
#splittinginto 2 different return valuse
#lambda <parameters>: <value_to_return_if_true> if <conditional> else <value_to_return_if_false>

evenorodd = lambda num 'Even' if num%2==0 else 'Odd'

print(evenorodd(1))
print(evenorodd(2))

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [15]:
f = lambda a: a**3

f(3)

27

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [16]:
letters = ['a', 'b', 'c', 'd']
capitalized = []
for a in letters: 
    capitalized.append(a.upper())
print(capitalized)


# same thiing but witha  map
capitalized = map(str.upper, letters)
print(capitalized)
capitalized = list(capitalized)
print(capitalized)

['A', 'B', 'C', 'D']
['A', 'B', 'C', 'D']


#### Using Lambda's with Map

In [1]:
def squareUnderTen(num1, num2):
    if num1<10 and num2<10:
        return num1**2, num2**2
    else:
        return num1, num2
    
lista = [4, 2, 20, 3, 5, 20]
listb = [4, 3, 10, 32, 6, 2]

squared_under_10 = []
for i in range(len(lista)):
    squared_under_10.append(squareUnderTen(lista[i], listb[i]))
    
print(squared_under_10)

#loop into map
#maps can accept multiple lists if the functiona ccepts multipe pieces of input
squared_under_10_map = list(map(squareUnderTen, lista, listb))
print(squared_under_10_map)


#rewrite the funciton as a lambda
lambda x,y: (x**2, y**2) if x<10 and y<10 else (x,y)

squaredu10 = list(map(lambda x,y: (x**2, y**2) if x<10 and y<10 else (x,y), lista, listb))
print(squaredu10)

[(16, 16), (4, 9), (20, 10), (3, 32), (25, 36), (20, 2)]
[(16, 16), (4, 9), (20, 10), (3, 32), (25, 36), (20, 2)]
[(16, 16), (4, 9), (20, 10), (3, 32), (25, 36), (20, 2)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [6]:
g = list(map(lambda x,y: ((x*2)-1,(y*2)-1), lista, listb)) 
print(g)

[(7, 7), (3, 5), (39, 19), (5, 63), (9, 11), (39, 3)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [7]:
#filter(function, list)
names = ['Bob','Andy','Max','Evan','Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
new_names_list = list(filter(a_names, names))
print(new_names_list)

print(a_names)

['Andy', 'Angelica']
<function a_names at 0x000002B0840E8160>


#### Using Lambda's with Filter()

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [18]:
import statistics
from random import randint
nums = [randint(0,100) for x in range(20)]
print(nums)
z = statistics.mean(nums)
answer = [x for x in nums if z < x ]
print(z)
print(answer)

answer2 = list(filter(lambda v: v > mean(nums), nums))


[8, 10, 22, 34, 13, 14, 83, 5, 29, 19, 98, 88, 72, 29, 32, 3, 24, 52, 99, 56]
39.5
[83, 98, 88, 72, 52, 99, 56]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [23]:
#reduce(function, iterable)
#return a single value that is a result of the reduction of the list

# the way the function is designed is to accept two inputs and return one output
# the classic example of reduce() is teh summation of a list or finding the max/min value in a list
from functools import reduce
summation=reduce(lambda a,b: a+b, [5,10,20,25,30,35])
print(summation)

maximum = reduce(lambda a,b: a if a>b else b,[5,10,20,25,30,35])
print(maximum)

125
35


#### Using Lambda's with Reduce()

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [26]:
my_list = [1, 2, 3, 4, 5]

from functools import reduce

summation = reduce(lambda a,b: a*b, my_list)
print(summation)

120


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [27]:
#rewriting some equation or process as depending on a modified version of itself
#gives you your recursive call
#i.e. f(n) can be rewritten here as n + f(n-1)
#and then finding your base case or endpoint for the same function
# in this scenario f(1) = 1
#the way this actually runs is we build a recursive chain of funciton
def summation(n):
    if n==1:
        return 1
    return n + summation(n-1)

print(summation(5))

15


#### Writing a Factorial Function

In [28]:
def factorial(n):
    if n== 1:
        print('factorial(1) = 1')
        return 1
    print(f'factorial({n}) = {n} * factorial({n-1})')
    return n*factorial(n-1)

factorial(6)

factorial(6) = 6 * factorial(5)
factorial(5) = 5 * factorial(4)
factorial(4) = 4 * factorial(3)
factorial(3) = 3 * factorial(2)
factorial(2) = 2 * factorial(1)
factorial(1) = 1


720

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [31]:
def factorial2(n):
    if n <= 1:
        print('factorial(1) = 1')
        return 1
    print(f'factorial({n}) = {n} - factorial({n-1})')
    return n - factorial2(n-1)

factorial2(5)



factorial(1) = 1
facorial(2) = 1
factorial (3) = 2
factorial (4) = 2
factorial(5) = 3

factorial(5) = 5 - factorial(4)
factorial(4) = 4 - factorial(3)
factorial(3) = 3 - factorial(2)
factorial(2) = 2 - factorial(1)
factorial(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [34]:
def my_square(start, stop, step=-1):
    while start >= stop:
        yield start**2
        start += step
        
for i in my_square(10,0):
    print(i)

100
81
64
49
36
25
16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [17]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
#filter(function, list)

def my_filter(place):
    while "" in places:
        places.remove("")
    if place[0].isalpha():
        return True
    else:
        return False
    
answer10 = list(filter(my_filter, places))
print(answer10)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [33]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

#lambda x,y: x.lower() y.upper(), 

for i in range(len(author)):
    new_author = author[i].title()
    author[i] = new_author

def names_split(name_list):
    sorted_author = []
    for i in range(len(name_list)):
        x = sorted_author.append(name_list[i].split())
    return sorted_author
       
my_authors = names_split(author)

final_list = sorted(my_authors, key=lambda x:x[-1])

for i in range(len(final_list)):
    final_list[i] = ' '.join(final_list[i])


print(final_list)

['Victor Anisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David Hasselhoff']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [51]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
#squaredu10 = list(map(lambda x,y: (x**2, y**2) if x<10 and y<10 else (x,y), lista, listb))
#map(function, list)

def calculation (location,celsius):
    f = (9/5)*celsius + 32
    return location, f

#answer = list(map(lambda x,y: (f"'{x}', {(9/5)* y + 32}"), places))
answer = list(map(lambda x: (f"'{x[0]}' {(9/5)* x[1] + 32}"), places))
print(answer)

["'Nashua' 89.6", "'Boston' 53.6", "'Los Angelos' 111.2", "'Miami' 84.2"]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`